### DISCUSSION TITLE
PicoVision Beta-testers Quick-start Guide

### DISCUSSION THREAD

**Discussion by Gadgetoid** [pimoroni/picovision#3] [Announcements]
Hey! If you're reading this, chances are you've just unboxed a brand new PicoVision. Welcome to the 1980s- except have HDMI, WiFi and Bluetooth and USB and MicroPython.

PicoVision is a wireless HDMI stick for putting pico-powered, network-connected graphics onto your TV or HDMI monitor.

It works by combining a Pico W (which we call the "CPU") with a separate RP2040 (which we call the "GPU") for handling the HDMI output. They communicate via a pair of physical buffers- PSRAM chips- which are swapped via a hardware mux. The idea is that the CPU prepares a frame, writes it to the buffer, requests a swap and gets back the other, empty buffer.

The GPU code is embedded into your Pico W program code and hot-loaded onto the RP2040 at runtime, so everything is mutable and - should you wish to push the envelope - you don't have to rely on our interpretation of how things work.

Much of PicoVision was made possible by @MichaelBell whose enthusiasm pushed the project forward, and whose code makes this balmy setup possible.

# C++

To write C++ programs for PicoVision you'll need Michael's `dv_stick` branch, which you can find here: https://github.com/MichaelBell/pimoroni-pico/tree/dv_stick

This includes -

* drivers for talking to the PSRAM chips - https://github.com/MichaelBell/pimoroni-pico/tree/dv_stick/drivers/aps6404
* a driver to talk to the "GPU" - https://github.com/MichaelBell/pimoroni-pico/tree/dv_stick/drivers/dv_display
* a hard copy of the "GPU" firmware - https://github.com/MichaelBell/pimoroni-pico/blob/dv_stick/drivers/dv_display/pico-stick.h
* some examples - https://github.com/MichaelBell/pimoroni-pico/tree/dv_stick/examples/dv_stick

If you're familiar with PicoGraphics you should be able to start pushing pixels within minutes, but there are some small caveats:

* Because we have a pair of buffers, you wont get the same buffer back when you "display.update()"
* The GPU supports scrolling and sprites, which are not concepts PicoGraphics is aware of- you can do some serious pixel pushing with these.

# MicroPython

We have a MicroPython firmware up and running on PicoVision that lets you use PicoGraphics - plus all the usual MicroPython niceties - to push pixels in a more forgiving environment.

The pace of development is fast, so you should be familiar with grabbing artifact builds from GitHub.

You can find firmware for various commits and pull-requests by going to Actions here: https://github.com/pimoroni/picovision/actions

And click a "MicroPython Firmware" build. Now you'll see a list of Artifacts at the bottom, you'll normally want the ~2MB build so you don't nuke your filesystem, eg: 
`pimoroni-picovision-069f0e0d6436de698498f6640b0cd162d4fb3ac1-micropython.uf2 ` from https://github.com/pimoroni/picovision/actions/runs/5525452531

MicroPython is chopping and changing as we figure out what shape the Pico Vision API should take- give us a shout for example code.

# GPU "Firmware"

The GPU firmware is available here - https://github.com/MichaelBell/pico-stick

This is where grabbing a frame from the PSRAM buffer and outputting it to HDMI - using @Wren6991's PicoDVI - happens.

The GPU is controlled via an i2c interface, allowing for sprite positions and scanline scrolling to be updated without a buffer swap.

# Discussion

Feel free to use these forums, GitHub issues or PRs to discuss the various aspects of Pico Vision.

We also have a channel on the Pimoroni discord for more casual discussion - https://discord.gg/knKhkwtE5

**Comment by Daft-Freak** [comment]
Is there a pinout somewhere?

**Reply by MichaelBell** [reply]
Not sure which version you have, but think this is correct for the black one.  I don't think there were any changes for the newer green one that says "PicoVision" on it, but I could be wrong!
https://docs.google.com/spreadsheets/d/15mo4HeZC-ilqELLFrKgFgvInpzVUKKJCa2kxyyi1gj8/edit#gid=0

**Reply by Daft-Freak** [reply]
Ah nice, the audio pins are the same as the VGA/DV boards.

**Comment by Daft-Freak** [comment]
I should really clean up and split this from all the other changes, but... 32blit-sdk port: https://github.com/Daft-Freak/32blit-beta/tree/picovision-and-everything-else

Possibly working artefacts here: https://github.com/Daft-Freak/32blit-beta/actions/runs/5568692843 (with hires == 320x240, `-DDISPLAY_WIDTH=640 -DDISPLAY_HIEGHT=480` also works for hires == 640x480, lores == 320x240)

EDIT: PicoVision only branch: https://github.com/Daft-Freak/32blit-beta/tree/picovision

**Reply by MichaelBell** [reply]
Awesome!  Just trying this now :)

**Reply by Gadgetoid** [reply]
No luck getting the artefacts working-

- stick all .blit files on a microSD card
- flash `blit-loader.uf2`?
- if I reset, pops back up ` RPI-RP2`

**Reply by MichaelBell** [reply]
I was loading the uf2s in `examples` directly and they work fine (though I don't have access to a USB HID device that works as a controller so I didn't get much further than that...)

**Reply by Gadgetoid** [reply]
Oh I'd gone the blit route to see the launcher working :laughing: 

I wonder if any of my controllers work...

**Reply by Gadgetoid** [reply]
Weird, the .uf2 files not only don't work for me, but if I press reset after flashing one it drops me back to the RPI-RP2 bootloader MSD again.

Flashed my MicroPython build and it was fine. Weird!

**Reply by Daft-Freak** [reply]
Yeah, I haven't got to trying that. Don't think there's support for loading things from the SD card yet. (...and you can't use the CDC loader as it's in host mode...)

Maybe should've mentioned that, but that artefact is called Pico**System**-blit (though since the PicoVision port doesn't use the framebuffer allocation, those blits would probably work if the loader did :thinking: )

**Reply by Daft-Freak** [reply]
Pushed a more reasonable PicoVision branch (only includes two open PRs!). Slightly broken in the middle due to being cherry-picked over some refactors, but seems to work in the end.

(Still includes some hacks, also if you're doing things with gamepads the everything branch has a useful modified hardware-test for that)

**Reply by Daft-Freak** [reply]
So the magic command to get a .blit info flash without the usual tool is:
`picotool load [game].blit -t bin -o 0x10080000`

... and with a [special launcher](https://github.com/pimoroni/picovision/files/12072229/launcher-256k.blit.zip)

(loaded with `picotool load launcher-256k.blit -t bin -o 0x10040000`)

You can have two things. Though it seems a bit broken right now. (Some examples immediately fault)

**Reply by Daft-Freak** [reply]
The launcher stuff should be more reasonable now, also wrote some stuff down in this gist: https://gist.github.com/Daft-Freak/a326d3d6e164d70f9fe25d1353278b4b.

(Should be able to flash the PV blit-loader and run the PS .blits from the SD card now, use the special launcher to keep it after running something)

Possibly going a bit off-topic :laughing:

**Reply by Gadgetoid** [reply]
Woohoo, after a flash nuke and actually following the instructions rather than trying things at random- I'm successfully launching games. This is just bonkers! Nice work.
